wurst/event/DamageEvent.wurst

```diff
@@ -0,0 +1,385 @@
+
+    construct(int id, unit source, unit target, real amount, DamageType damageType, DamageElement damageElement)
+        this.id = id 
+        this.source = source 
+        this.target = target 
+        this.amount = amount 
+        this.originalAmount = amount 
+        this.damageType = damageType
+        this.damageElement = damageElement
+
+        count++ 
+        stack[count] = this 
+        current = this 
+
+    ondestroy
+        count-- 
+        current = stack[count]
+
+
+/* DAMAGE EVENT */
+
+public class DamageEvent 
+
+    // Next damage instance
+    protected static int            nextDamageId      = 0 
+    protected static DamageType     nextDamageType    = UNKNOWN 
+    protected static DamageElement  nextDamageElement = null 
+    // Mid damage instance
+    protected static bool abort = false 
+    // Damage listeners
+    protected static DamageListener array firstListeners
+    protected static int maxPriority = 0 
+    
+
+    /* ON DAMAGE */
+    protected static function onDamage()
+        var amount = GetEventDamage()
+        if amount == 0. 
+            return 
+        if nextDamageType == DamageType.UNKNOWN 
+            // Damage type hasn't been defined -> detect it
+            if DETECT_NATIVE_ABILITIES 
+                if amount < 0.
+                    nextDamageType = SPELL
+                    amount *= -1
+                else
+                    nextDamageType = ATTACK
+                    nextDamageElement = DAMAGE_ELEMENT_ATTACK 
+            else 
+                nextDamageType = ATTACK
+                nextDamageElement = DAMAGE_ELEMENT_ATTACK 
+
+        let dmg = new DamageInstance(nextDamageId, GetEventDamageSource(), GetTriggerUnit(), amount, nextDamageType, nextDamageElement)
+        
+        nextDamageId = 0 
+        nextDamageType = UNKNOWN
+        nextDamageElement = null
+        
+        for i = 0 to maxPriority
+            var listener = firstListeners[i]
+            while listener != null
+                listener.onEvent()
+                if abort                    
+                    dmg.amount = 0
+                    break
+                else 
+                    listener = listener.next
+            if abort 
+                break 
+
+        BlzSetEventDamage(dmg.amount)
+
+        destroy dmg
+        abort = false  
+
+        
+    /* LISTENERS */
+    /** Add a damage event listener.
+        If The order of firing is important, use addListener(priority, listener) */
+    static function addListener(DamageListener listener) returns DamageListener
+        return addListener(maxPriority, listener)
+
+    /** Add a damage event listener with a given priority.
+        Listeners of different priorities fire from the lowest priority to the highest priority added. 
+        Listeners of the same priority fire by order of addition (FIFO) */
+    static function addListener(int priority, DamageListener listener) returns DamageListener
```

Comment:
@WareditorGit 

Thanks for this. Your library is now the only one that properly handles 0-damage-events in 1.31.x (and I don't plan to fix DamageType or DamageDetection)

These docs need to describe constraints for priority. What is the max value, and what is the minimum one? The first thing I tried to do when using the priority management is set `-1`, which obviously won't work.